backend/chunker/chunker.go

```diff
@@ -0,0 +1,1412 @@
+			remote := object.Remote()
+			if isSubdir[remote] {
+				if hardErrors {
+					return nil, fmt.Errorf("%q is both meta object and directory", remote)
+				}
+				badEntry[remote] = true // fall thru
+			}
+			if badEntry[remote] {
+				fs.Debugf(f, "invalid directory entry %q", remote)
+				continue
+			}
+			if err := object.validate(); err != nil {
+				if hardErrors {
+					return nil, err
+				}
+				fs.Debugf(f, "invalid chunks in object %q", remote)
+				continue
+			}
+		}
+		chunkedEntries = append(chunkedEntries, entry)
+	}
+
+	return chunkedEntries, nil
+}
+
+// NewObject finds the Object at remote.
+func (f *Fs) NewObject(ctx context.Context, remote string) (fs.Object, error) {
+	if f.nameRegex.MatchString(remote) {
+		return nil, fmt.Errorf("%q should be meta object, not a chunk", remote)
+	}
+
+	var (
+		object *Object
+		rawObj fs.Object
+		err    error
+	)
+
+	if f.useMeta {
+		rawObj, err = f.Fs.NewObject(ctx, remote)
+		if err != nil {
+			return nil, err
+		}
+		remote = rawObj.Remote()
+
+		// meta object cannot be large
+		object = f.newObject("", rawObj, nil)
+		if object.size > maxMetaDataSize {
+			return object, nil
+		}
+	} else {
+		object = f.newObject(remote, nil, nil)
+	}
+
+	// the object is small, probably it contains meta data
+	dir := path.Dir(strings.TrimRight(remote, "/"))
+	if dir == "." {
+		dir = ""
+	}
+	entries, err := f.Fs.List(ctx, dir)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot detect chunked object")
+	}
+
+	for _, dirOrObject := range entries {
+		entry, ok := dirOrObject.(fs.Object)
+		if !ok {
+			continue
+		}
+		entryRemote := entry.Remote()
+		if !strings.HasPrefix(entryRemote, remote) {
+			continue
+		}
+		match := f.nameRegex.FindStringSubmatchIndex(entryRemote)
+		if match == nil || entryRemote[match[2]:match[3]] != remote {
+			continue
+		}
+		chunkNum, _ := strconv.Atoi(entryRemote[match[4]:match[5]])
+		//fs.Debugf(f, "%q belongs to %q as chunk %d", entryRemote, remote, chunkNum)
+		if err := object.addChunk(entry, chunkNum); err != nil {
+			return nil, err
+		}
+	}
+
+	if object.main == nil && (object.chunks == nil || len(object.chunks) == 0) {
+		return nil, fs.ErrorObjectNotFound
+	}
+	// calculate total object size
+	if err := object.validate(); err != nil {
+		return nil, err
+	}
+	return object, nil
+}
+
+// put implements Put, PutStream, PutUnchecked, Update
+func (f *Fs) put(ctx context.Context, in io.Reader, src fs.ObjectInfo, remote string, options []fs.OpenOption, rawPut putFn) (fs.Object, error) {
+	// Unwrap the accounting
+	unwrapIn, wrapBack := accounting.UnWrap(in)
+
+	c := &chunkingReader{
+		rawReader: unwrapIn,
+		readCount: 0,
+		chunkSize: int64(f.opt.ChunkSize),
+		sizeTotal: src.Size(),
+	}
+	c.chunkLimit = c.chunkSize
+	c.sizeLeft = c.sizeTotal
+
+	// Wrap the accounting back on
+	wrapIn := wrapBack(c)
+
+	baseRemote := remote
+	tempRemote := fmt.Sprintf("%s__tmp.%d.", baseRemote, time.Now().Unix())
+
+	// Prepare to rename
```

Comment:
It would be better to use `operations.Move` here as if the underlying remote doesn't support Move it will fallback to Copy+Delete.  None of the bucket based remotes support Move.  Nearly every rclone backend supports Move or Copy now.

Use `operations.CanServerSideMove` to check that `operations.Move` doesn't do a download then re-upload.